bundles/automation/org.eclipse.smarthome.automation.providers/src/main/java/org/eclipse/smarthome/automation/core/provider/AutomationResourceBundlesEventQueue.java

```diff
@@ -0,0 +1,196 @@
+                }
+                l_queue = queue;
+                shared = true;
+            }
+            try {
+                Iterator events = l_queue.iterator();
+                while (events.hasNext()) {
+                    processBundleChanged((BundleEvent) events.next());
+                }
+            } catch (Throwable t) {
+                t.printStackTrace(); // TODO
+            }
+            synchronized (this) {
+                if (shared)
+                    queue.clear();
+                shared = false;
+                waitForEvents = true;
+                notifyAll();
+            }
+        }
+    }
+
+    public void stop() {
+        synchronized (this) {
+            closed = true;
+            notifyAll();
+        }
+        bTracker.close();
+    }
+
+    /**
+     * @see org.osgi.util.tracker.BundleTrackerCustomizer#addingBundle(org.osgi.framework.Bundle,
+     *      org.osgi.framework.BundleEvent)
+     */
+    @Override
+    public Object addingBundle(Bundle bundle, BundleEvent event) {
+        if (isAnAutomationProvider(bundle)) {
+            addEvent(event == null ? new BundleEvent(BundleEvent.INSTALLED, bundle) : event);
+        }
+        return bundle;
+    }
+
+    /**
+     * @see org.osgi.util.tracker.BundleTrackerCustomizer#modifiedBundle(org.osgi.framework.Bundle,
+     *      org.osgi.framework.BundleEvent, java.lang.Object)
+     */
+    @Override
+    public void modifiedBundle(Bundle bundle, BundleEvent event, Object object) {
+        if (isAnAutomationProvider(bundle) && event.getType() == BundleEvent.UPDATED) {
+            addEvent(event);
+        }
+    }
+
+    /**
+     * @see org.osgi.util.tracker.BundleTrackerCustomizer#removedBundle(org.osgi.framework.Bundle,
+     *      org.osgi.framework.BundleEvent, java.lang.Object)
+     */
+    @Override
+    public void removedBundle(Bundle bundle, BundleEvent event, Object object) {
+        if (!closed) {
+            if (isAnAutomationProvider(bundle)) {
+                addEvent(event);
+            }
+        }
+    }
+
+    /**
+     * This method is used to check if the specified {@link Bundle} contains resource files providing {@link Template}s.
+     *
+     * @param bundle is a {@link Bundle} object to check.
+     * @return <tt>true</tt> if the specified {@link Bundle} contains resource files providing {@link Template}s,
+     *         <tt>false</tt> otherwise.
+     */
+    private boolean isAnAutomationProvider(Bundle bundle) {
+        return bundle.getHeaders().get(AUTOMATION_RESOURCES_HEADER) != null;
+    }
+
+    private synchronized void addEvent(BundleEvent event) {
+        if (closed)
+            return;
+        if (shared) {
+            queue = new ArrayList();
+            shared = false;
+        }
+        if (queue.add(event)) {
+            if (running)
+                notifyAll();
+            else {
+                Thread th = new Thread(this, "Template Provider Processing Queue");
```

Comment:
It's not only for Template providers, right? So better change to "Automation Provider"